fx-src/FirefoxStorage.js

```diff
@@ -0,0 +1,227 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import BaseAdapter from "../src/adapters/base";
+
+Components.utils.import("resource://gre/modules/Sqlite.jsm");
+Components.utils.import("resource://gre/modules/Task.jsm");
+
+const statements = {
+  "createCollectionData": `
+    CREATE TABLE collection_data (
+      collection_name TEXT,
+      record_id TEXT,
+      record TEXT
+    );`,
+
+  "createCollectionMetadata": `
+    CREATE TABLE collection_metadata (
+      collection_name TEXT PRIMARY KEY,
+      last_modified INTEGER
+    ) WITHOUT ROWID;`,
+
+  "createCollectionDataRecordIdIndex": `
+    CREATE UNIQUE INDEX unique_collection_record
+      ON collection_data(collection_name, record_id);`,
+
+  "clearData": `
+    DELETE FROM collection_data
+      WHERE collection_name = :collection_name;`,
+
+  "createData": `
+    INSERT INTO collection_data (collection_name, record_id, record)
+      VALUES (:collection_name, :record_id, :record)`,
+
+  "updateData": `
+    UPDATE collection_data
+      SET record = :record
+        WHERE collection_name = :collection_name
+        AND record_id = :record_id`,
+
+  "deleteData": `
+    DELETE FROM collection_data
+      WHERE collection_name = :collection_name
+      AND record_id = :record_id`,
+
+  "saveLastModified": `
+    REPLACE INTO collection_metadata (collection_name, last_modified)
+      VALUES (:collection_name, :last_modified)`,
+
+  "getLastModified": `
+    SELECT last_modified
+      FROM collection_metadata
+        WHERE collection_name = :collection_name`,
+
+  "getRecord": `
+    SELECT record
+      FROM collection_data
+        WHERE collection_name = :collection_name
+        AND record_id = :record_id`,
+
+  "listRecords": `
+    SELECT record
+      FROM collection_data
+        WHERE collection_name = :collection_name`
+};
+
+const createStatements = ["createCollectionData",
+                          "createCollectionMetadata",
+                          "createCollectionDataRecordIdIndex"];
+
+export default class FirefoxAdapter extends BaseAdapter {
+  constructor(dbname) {
+    super();
+    this.dbname = dbname;
+
+    // make a promise to initialize
+    const self = this;
+    this.init = Task.spawn(function* () {
+      let connection;
+      try {
+        connection = yield self._doOpenConnection();
+        yield connection.executeTransaction(function* doSetup() {
+          let schema = yield (connection.getSchemaVersion());
+
+          if (schema == 0) {
+
+            for (let statementName of createStatements) {
+              yield (connection.execute(statements[statementName]));
+            }
+
+            yield (connection.setSchemaVersion(1));
+          } else if (schema != 1) {
+            throw new Error("Unknown database schema: " + schema);
+          }
+        });
+        self.initialized = true;
```

Comment:
question on your first line comment here, @rnewman; I'm not sure what you mean? Should I also be manually checking the schema each time I initialize? Surely, since for a given adapter instance it will only open a connection for a single file with a similar set of params, we should be OK here?